Subscriptions: Add unit tests and fixtures to validate block content parsing

ramonjd opened the pull request:
### Changes proposed in this Pull Request:
This PR adds unit and block fixture parsing tests to the Subscriptions block, *and* extricate the controls into a `controls.js` file. 

No block functionality is affected.

### Jetpack product discussion
p1HpG7-b3G-p2

###  Does this pull request change what data or activity we track or use?
No.

### Testing instructions:

Running the unit tests:

`cd projects/plugins/jetpack && yarn fixtures:test extensions/blocks/subscriptions/test/`

We should also ensure that the block works as expected in the editor and saved stats on a site.

To test the subscriptions block in a browser we need an active connection or have the plugin running on WPCOM in order to activate the module. 

Apply D57958-code to your sandbox and add a subscriptions block to a new sandboxed test site.

<img width="832" alt="Screen Shot 2021-03-04 at 11 53 21 am" src="https://user-images.githubusercontent.com/6458278/109895392-401d1400-7ce3-11eb-85d7-92b0cb432931.png">
<img width="735" alt="Screen Shot 2021-03-04 at 12 15 40 pm" src="https://user-images.githubusercontent.com/6458278/109895483-6347c380-7ce3-11eb-9e24-6eb88d0cb89d.png">


### Proposed changelog entry for your changes:
None. No functional changes made.

ramonjd committed:
[not verified] Adding initial fixture files

ramonjd committed:
[not verified] Reverting test code

ramonjd committed:
Extricating inspector controls children into an external component and adding initial control tests.

ramonjd committed:
[not verified] Adding more control tests

ramonjd committed:
[not verified] Adding edit tests

andrewserong commented on projects/plugins/jetpack/extensions/blocks/subscriptions/test/controls.js:
```diff
@@ -0,0 +1,203 @@
+		test( 'set space inside', () => {
+			render( <SubscriptionsInspectorControls { ...defaultProps } /> );
+			userEvent.click( screen.getByText( 'Spacing Settings' ), { selector: 'button' } );
+			// For some reason userEvent.type won't work on these inputs
```
Looks like this and the below one might not work with `userEvent.type` because the default value is double-digits and the max value for the input field is also double digits. If I change the the props we render the `SubscriptionsInspectorControls` with in the test to start with a `padding` of `0`, then I can get it to work:

```
			const propsWithPaddingOverride = { ...defaultProps, padding: 0 };
			render( <SubscriptionsInspectorControls { ...propsWithPaddingOverride } /> );
			userEvent.click( screen.getByText( 'Spacing Settings' ), { selector: 'button' } );
			userEvent.type( screen.getAllByLabelText( 'Space Inside' )[1], '13' );

			expect( setAttributes ).toHaveBeenLastCalledWith( {
				padding: 13,
			} );
```

I think it's fine to stick with `userEvent.clear` to test that it calls `setAttributes`, I just wanted to work out why it wasn't working 🙂

ramonjd replied:
Nice. Fresh 👀. Thanks!

andrewserong reviewed (commented):
This is testing well for me @ramonjd, I tried the block out in a sandboxed wpcom environment and in my local Jetpack site, and the block still works as expected.

I left a tiny comment about the padding tests, but I also noticed with the `edit.js` tests it appears to be trying to fire the `apiFetch`. It doesn't seem to be breaking the tests, but does log out a warning when I run `yarn jest extensions/blocks/subscriptions/test/`:

![image](https://user-images.githubusercontent.com/14988353/110061919-7aa6af80-7dbc-11eb-8b34-790890d8b5df.png)

Is it worth mocking `apiFetch` for the Edit tests?

ramonjd commented on projects/plugins/jetpack/extensions/blocks/subscriptions/controls.js:
```diff
@@ -0,0 +1,247 @@
+	DEFAULT_FONTSIZE_VALUE,
+} from './constants';
+
+export default ( {
```
Yeah for sure. More clarity is better!

ramonjd commented:
> Is it worth mocking apiFetch for the Edit tests?

Thanks for pointing this out @andrewserong! Totally agree we should remove those warnings.

It also gave me a chance to learn about [act()](https://reactjs.org/docs/test-utils.html#act)

ramonjd committed:
[not verified] Naming the anonymous control function to SubscriptionControls
Adding mocks for apiFetch in the edit tests
Using zero values for default input[type=number] fields to receive more reliable userEvent.type results

glendaviesnz commented:
This tested well for me, and api notices are no longer logged.

For some reason the block content migrates every time you load a page in the editor with this block, even though the save function and saved content are the same:

```
New content generated by `save` function:

<div class="wp-block-jetpack-subscriptions wp-block-jetpack-subscriptions__supports-newline wp-block-jetpack-subscriptions__use-newline wp-block-jetpack-subscriptions__show-subs">[jetpack_subscription_form show_subscribers_total="true" button_on_newline="true" custom_font_size="36px" custom_border_radius="19" custom_border_weight="8" custom_border_color="#cf2e2e" custom_padding="30" custom_spacing="44" submit_button_classes="has-36-px-font-size has-cf-2-e-2-e-border-color has-text-color has-black-color has-background has-light-green-cyan-background-color" email_field_classes="has-36-px-font-size has-cf-2-e-2-e-border-color" show_only_email_and_button="true"]</div>

Content retrieved from post body:

<div class="wp-block-jetpack-subscriptions wp-block-jetpack-subscriptions__supports-newline wp-block-jetpack-subscriptions__use-newline wp-block-jetpack-subscriptions__show-subs">[jetpack_subscription_form show_subscribers_total="true" button_on_newline="true" custom_font_size="36px" custom_border_radius="19" custom_border_weight="8" custom_border_color="#cf2e2e" custom_padding="30" custom_spacing="44" submit_button_classes="has-36-px-font-size has-cf-2-e-2-e-border-color has-text-color has-black-color has-background has-light-green-cyan-background-color" email_field_classes="has-36-px-font-size has-cf-2-e-2-e-border-color" show_only_email_and_button="true"]</div>
```
I can't see an obvious reason why this is happening, but it looks like this is an existing bug as it happens on WPcom currently, so not sure if you want to investigate further or add a separate issue for it.

aaronrobertshaw commented:
This tests well for me 👍 

> For some reason the block content migrates every time you load a page in the editor with this block, even though the save function and saved content are the same:

With empty/default attributes the current block version's shortcode matches the v1 deprecation. That deprecation implements `isEligible` which checks if a `subscribeButton` attribute is empty or not. It determines the deprecation to be eligible if that attribute is empty. 

As the current default subscribe block content matches and is "valid", the `isEligible` is run. Given the attributes won't include `subscribeButton`, that explains why it is being mistakenly run.

If that `isEligible` [function](https://github.com/Automattic/jetpack/blob/add/subscription-block-tests/projects/plugins/jetpack/extensions/blocks/subscriptions/deprecated/v1/index.js#L24) is deleted, the block deprecation doesn't run. 

It appears that the point of this deprecation was to migrate from the `subscribeButton` attribute to `submitButtonText`. Instead of deleting `isEligible` I think updating the check to see if the attributes contain the `subscribeButton` key should still allow the deprecation to cover its intended use case. 

```js
	isEligible: attr => {
		if ( ! attr.hasOwnProperty( 'subscribeButton' ) || ! isEmpty( attr.subscribeButton ) ) {
			return false;
		}
		return true;
	},
```

Making a change like this might give an opportunity to run the fixture tests in earnest as well? 🙂

ramonjd committed:
[not verified] deprecation v1 migrated from the subscribeButton attribute to submitButtonText.
To determine whether deprecation v1 is eligible we need to check for both the old property subscribeButton and its value since only checking for empty matches empty/default on other versions.

See: #18993#issuecomment-793490858

ramonjd committed:
[not verified] Update deprecation v2 with valid attributes

ramonjd committed:
[not verified] Add changelog

ramonjd committed:
[not verified] Removing subsequent-version's classname in v2 HTML attributes

aaronrobertshaw committed:
[not verified] # This is a combination of 3 commits.

 Bring all deprecations and fixtures into line

aaronrobertshaw commented:
I've pushed a commit updating all the fixtures and some of the deprecations. All the tests are green for me and all the deprecations' `.json` files show `isValid: true`.

Ran the tests as per the PR instructions but also via `yarn jest extensions/blocks/subscriptions/test/` to execute only this block's tests.

There are a couple of things to note.
- Deprecation v4 was added to be able to migrate from shortcode that had localized default text strings embedded in it. I wasn't sure if I could set a locale for the fixture test so that is something that could be improved.
- This block will need testing in the post editor again to ensure I haven't caused a regression.
- I added a brief explanation as to what I understand the reasoning for each deprecation to be. Feel free to change those comments' wording or structure. Their only aim is to help our future selves and others.

ramonjd commented:
Needs rebase after https://github.com/Automattic/jetpack/commit/d001afc08e7a70209cfdaafbbca14a91b0e65a25

ramonjd committed:
[not verified] Remove unused dependency
Clearing some controls before modifying values

jeherve commented:
> I suspect we need to update our dependency on @wordpress/components to get the latest UnitControl changes
>
> We're pulling 9.2.6, but the latest is 12.0.7

Noting that this was planned in #16763, but we've run into some issues that have stopped from updating so far. :(

ramonjd committed:
[not verified] We need to mock the button width control component temporarily to avoid type/import errors that should be fixed when we update the wordpress monorepo

ramonjd commented:
I've mocked the button width control component (similar to the [Button tests](https://github.com/Automattic/jetpack/pull/19002/commits/0edac7df052558478bd59f24ca75f21f7a1c4f8d)) in the meantime to move this PR forward.

stacimc reviewed (approved):
Lgtm, tests are passing now 👍

ramonjd commented:
This patch uses `ButtonWidthControl` which was included in https://github.com/Automattic/jetpack/pull/18809

To keep things in sync https://github.com/Automattic/jetpack/pull/18809 needs to be deployed to WPCOM before we deploy this patch to WPCOM (and after the next Jetpack release)

jeherve commented:
r223841-wpcom
